### PR TITLE
refactor(storage): use enum for object store impl

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Currently in our code, we will add a prefix in the object store path to indicate whether we are operating on local or remote object store, and therefore all our object store implementation (S3ObjectStore, InMemObjectStore, DiskObjectStore ...) has to internally be aware that whether they are local object store or not, which makes the logic wired and messy. 

In this PR, we use an enum to represent `ObjectStoreImpl` and explicitly restrict which object store can be used as remote or local object store. The path check and routing is done in this higher level enum so that the real object store implementation doesn't have to be aware that whether they are local or not. The enums look like
```
pub enum ObjectStoreImpl {
    InMem(MonitoredObjectStore<InMemObjectStore>),
    Disk(MonitoredObjectStore<DiskObjectStore>),
    S3(MonitoredObjectStore<S3ObjectStore>),
    Hybrid {
        local: Box<ObjectStoreImpl>,
        remote: Box<ObjectStoreImpl>,
    },
}
```
Some macros are used to reduce duplicated code. 

Another benefit of introducing the enum is that in the code where use ObjectStoreImpl, we can be aware which type of object store we are using, and we may call some non-object-store-trait methods of specific object stores to get better performance in some scenario.

After the enum is introduced, we can also avoid using async_trait in object store, though the original cost of using async_trait can be ignored compared to external IO.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Comment by Little-Wallace** [comment]
It's not necessary to use enum for object store. Because the time cost on IO and RPC will be much more than which cost on dynamic virtual-table.

**Comment by wenym1** [comment]
> It's not necessary to use enum for object store. Because the time cost on IO and RPC will be much more than which cost on dynamic virtual-table.

Performance is not the main reason for introducing enum for object store in this PR. The reason is described in the PR description.

**Comment by skyzh** [comment]
Generally looks good. For @Little-Wallace 's concern, we can still add `async_trait` over the traits and impls, so that they still use dynamic dispatch even if they're in an enum.

File: src/object_store/src/object/mod.rs [hunk 1]
```diff
@@ -72,101 +91,199 @@ impl BlockLocation {
...
-    async fn readv(&self, path: &str, block_locs: &[BlockLocation]) -> ObjectResult<Vec<Bytes>> {
-        if is_local_path(path) {
-            self.local.readv(path, block_locs).await
-        } else {
-            self.remote.readv(path, block_locs).await
-        }
-    }
+pub enum RemoteObjectStore {
+    InMem(MonitoredObjectStore<InMemObjectStore>),
+    Disk(MonitoredObjectStore<DiskObjectStore>),
```
Comment (hzxa21) [inline src/object_store/src/object/mod.rs:216]: Why can we have InMem and Disk object store as remote object store?

Comment (wenym1) [reply src/object_store/src/object/mod.rs:216]: InMem is used as remote object store in unit test and e2e test. Disk might be used as remote object store when we want to deploy a simple system without S3 or MinIO, but want to use more storage space.

**Comment by Little-Wallace** [comment]
> Generally looks good. For @Little-Wallace 's concern, we can still add `async_trait` over the traits and impls, so that they still use dynamic dispatch even if they're in an enum, and thus avoid codegen explosion.

I just think it is not necessary to maintain a complex code to optimize performance of object-storage acquirement. Because every times we send a request to object storage we will switch thread and kernel-state and wait several milliseconds or several hundred milliseconds (because s3 is much slower than MinIO based on NVMe)

**Review by hzxa21 — APPROVED** [review]
The logics are greatly simplified. Thanks for the PR. LGTM!

Looks like the spot instance limit is hit in CI. You need to push an empty commit to rerun it.

**Pull request merged**